okhttp/src/main/java/io/grpc/okhttp/OkHttpClientTransport.java

```diff
@@ -955,5 +984,49 @@ public void alternateService(int streamId, String origin, ByteString protocol, S
         int port, long maxAge) {
       // TODO(madongfly): Deal with alternateService propagation
     }
+
+    /**
+     * This is called exactly once.
+     */
+    private void startDeadlineWatchdog(final Deadline deadline) {
+      Runnable onTimeout = new Runnable() {
+        @Override
+        public void run() {
+          if (firstSettings) {
+            onException(new IOException("ssl handshake timeout or frame settings timeout"));
+          }
+        }
+      };
+      synchronized (lock) {
+        if (watchdogCanceled != null) {
+          // watchdog canceled already before it even started
+          Preconditions.checkState(watchdogCanceled, "watchdogCanceled");
```

Comment:
There is no way that cancelDeadlineWatchdog could happen before startDeadlineWatchdog, since cancelDeadlineWatchdog is only called when we receive the HTTP/2 settings and startDeadlineWatchdog is called before frameWriter.becomeConnected is called (thus no HTTP/2 is going on when the watchdog is started). That also means you don't need the synchronization.